### ISSUE TITLE
PyOpenSSL is not working in ansible version 2.5 to 2.7

### ISSUE THREAD

**Issue by NiveditaKumari14** [ansible/ansible#66558]
PyOpenSSL giving an error in ansible version 2.5 to 2.7 ["changed": false, "msg": "You need to have PyOpenSSL>=0.15"]

##### SUMMARY
When pyopenssl Version: 19.1.0 is installed and running a playbook to generate an OpenSSL Certificate, an error is received that ""msg": "You need to have PyOpenSSL>=0.15""
[Note: Same behaviour is observed for ansible version 2.6 and 2.7]

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
openssl_csr

##### ANSIBLE VERSION
ansible 2.5.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible-2.5.3-py2.7.egg/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, May 20 2019, 12:21:26) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]

Note: Same behavior is observed in ansible version 2.6 and 2.7

##### OS / ENVIRONMENT
pip --version:  pip 19.3.1 from /usr/lib/python2.7/site-packages/pip (python 2.7)

pip show pyopenssl:
Name: pyOpenSSL
Version: 19.1.0
Summary: Python wrapper module around the OpenSSL library
Home-page: https://pyopenssl.org/
Author: The pyOpenSSL developers
Author-email: cryptography-dev@python.org
License: Apache License, Version 2.0
Location: /usr/lib64/python2.7/site-packages
Requires: six, cryptography
Required-by: insights-client

##### EXPECTED RESULTS
When pyopenssl is installed, then it should not give an error same as in ansible version below 2.5 and above 2.7

##### ACTUAL RESULTS
Giving an error "msg": "You need to have PyOpenSSL>=0.15" when pyopenssl latest version is installed

Remarks: Below version 2.5 and above version 2.8, pyopenssl is working correctly

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/crypto/openssl_csr.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/openssl_csr.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @MarkusTeufelberger @Shaps @Spredzy @Xyon @felixfontein @puiterwijk @resmo
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by felixfontein** [comment]
Please note that Ansible 2.7 no longer receives bugfixes, and Ansible 2.8 only receives critical bugfixes (a fix for this one probably won't count); see https://docs.ansible.com/ansible/devel/reference_appendices/release_and_maintenance.html#release-status So I'm not sure that it makes sense working on this since it seems to work with Ansible 2.9 according to you.

But about your issue: are you sure that pyOpenSSL is actually installed on the remote machine (where the module is executed) and installed for the same Python version Ansible is using on that machine? If you are not sure, you can use https://docs.ansible.com/ansible/devel/modules/python_requirements_info_module.html to make sure that pyOpenSSL is installed where it is used.

Finally, note that at least for the macOS integration test, pyOpenSSL 19.1.0 is used, and that seems to work fine (see https://app.shippable.com/github/ansible/ansible/runs/156246/27/console for an OSX test run with pyOpenSSL 19.1.0, which included the tests for openssl_xxx on the stable-2.7 branch). So this is more likely a problem with your configuration.

**Comment by NiveditaKumari14** [comment]
As Ansible 2.8 receives critical bugfixe, so the same issue is observed in ansible version 2.8. This problem is resolved in ansible 2.9 so, the code should be back ported to resolve pyopenssl issue in 2.8. 

>>are you sure that pyOpenSSL is actually installed on the remote machine (where the module is executed) and installed for the same Python version Ansible is using on that machine?

Yes, pyOpenSSL is installed on the remote machine and installed for the same Python version Ansible is using on that machine. 

AS my deployment environment is set on ansible version 2.5.3, so is there any workaround to solve this issue?

**Comment by felixfontein** [comment]
> As Ansible 2.8 receives critical bugfixe, so the same issue is observed in ansible version 2.8. This problem is resolved in ansible 2.9 so, the code should be back ported to resolve pyopenssl issue in 2.8.

Problems with modules are usually not considered critical. So the chances of a fix being backported are essentially nil. Also, the usage of PyOpenSSL between Ansible 2.8 and 2.9 did not change at all. What did change is that most openssl_* modules also support the cryptography library and prefer it if possible. You explicitly have to tell the module to use PyOpenSSL so that PyOpenSSL is actually used.

So it could well be that in your case, PyOpenSSL also doesn't work with Ansible 2.9, but you didn't notice since it used cryptography instead.

> > > are you sure that pyOpenSSL is actually installed on the remote machine (where the module is executed) and installed for the same Python version Ansible is using on that machine?
> 
> Yes, pyOpenSSL is installed on the remote machine and installed for the same Python version Ansible is using on that machine.
> 
> AS my deployment environment is set on ansible version 2.5.3, so is there any workaround to solve this issue?

You should stop using Ansible 2.5 since it is unmaintained. Upgrade to Ansible 2.9, since that does seem to solve your problems as well and is the current release.

**Comment by MarkusTeufelberger** [comment]
Not much that can be done here, this specific problem looks like user error, not an error in the openssl_* modules.

wontfix
